app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\InventorySales\Observer\Product;
+
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Class IsProductSalable
+ *
+ * @package Magento_InventorySales
+ */
+class IsProductSalable implements ObserverInterface
+{
+    /**
+     * @var IsProductSalableInterface
+     */
+    private $isProductSalable;
+
+    /**
+     * @var StockResolverInterface
+     */
+    private $stockResolver;
+
+    /**
+     * @var StoreManagerInterface
+     */
+    private $storeManager;
+
+    /**
+     * @var array
+     */
+    private $stockIdByWebsite = [];
+
+    /**
+     * IsProductSalable constructor.
+     *
+     * @param IsProductSalableInterface $isProductSalable
+     * @param StockResolverInterface $stockResolver
+     * @param StoreManagerInterface $storeManager
+     */
+    public function __construct(
+        IsProductSalableInterface $isProductSalable,
+        StockResolverInterface $stockResolver,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->isProductSalable = $isProductSalable;
+        $this->stockResolver = $stockResolver;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * Check if product is salable via MSI api
+     *
+     * @param Observer $observer
+     * @return void
+     */
+    public function execute(Observer $observer)
+    {
+        /** @var \Magento\Catalog\Model\Product $product */
+        $product = $observer->getProduct();
+        /** @var \Magento\Framework\DataObject $salableObject */
+        $salableObject = $observer->getSalable();
+
+        $websiteId = $product->getStore()->getWebsiteId();
+
+        $isSalable = $this->isProductSalable->execute($product->getSku(), $this->getStockId($websiteId));
+        $salableObject->setIsSalable($isSalable);
+    }
+
+    /**
+     * Retrieve stock_id by website
+     *
+     * @param int $websiteId The website Id
+     * @return int
+     */
+    private function getStockId($websiteId)
+    {
+        if (!isset($this->stockIdByWebsite[$websiteId])) {
```

Comment:
It would be great to avoid introducing in-object cache as it creates technical challenge for anyone attempting to run Magento in multi-threaded environment (for example some sort of solution like [this](https://github.com/appserver-io/appserver)).

What we can do is to introduce a caching facade later on which may be easily disabled with the di configuration.

At this point I would consider this cache to be a slightly premature optimization.